File: ip_scan.py
import socket
import ipaddress
import subprocess
import struct
import sys

SKIP_LIST = [int(ipaddress.IPv4Address("0.0.0.0")),
		range(int(ipaddress.IPv4Address("192.168.0.1")),int(ipaddress.IPv4Address("192.168.255.255"))),
		int(ipaddress.IPv4Address("127.0.0.1"))]


def make_dns_packet():
	'''
	make a customize udp packet (for dns query, carries a message to
	system admins)
	'''
	message = "For a Research @Case Western Reserve, Plz Contact jxm959@case.edu"
	message = bytes(message, "ASCII")


	transaction_id = 0xffff # 2 byte short
	control 	   = 0x0100
	q_counts       = 0x0001
	ans_counts     = 0x0000
	auth_counts    = 0x0000
	add_counts     = 0x0000 # 2 byte short
	query          = 0x0c
	type_          = 0x0001 # 2 byte short (A)
	class_         = 0x0001 # 2 byte short (IN)

	query_website  = bytes("case.edu", "ASCII")

	format_= "!hh4h"





def make_datagram_sockets() -> (socket.socket, socket.socket):
	'''
	make a pair of udp socket with our options
	1. send dns queries
	2. listen responses. 
	'''
	send_socket_instance = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) # Produce a UDP DGRAM socket
	recv_socket_instance = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP) # Produce a UDP DGRAM socket
	# send_socket_instance.setsockopt(socket.IPPROTO_IP, socket.IP_TTL,     255) # Set the TTL to 255

	return send_socket_instance, recv_socket_instance




def listen_and_check_response(in_socket: socket.socket, out_socket: socket.socket, packet, ip_address:ip_address.IPv4Address) -> bool:
	'''
	listen to the server side response and check whether it is a legal &
	valid response
	'''
	pass

def multiprocess_scan(start_ip:int, end_ip:int, process_num:int) -> None:
	'''
	scan the IPv4 address space from [start_ip] to [end_ip]
	with [process_num] of processes simultaneously
	'''
	pass


#====================== Private Methods ====================================

def _check_skip_policy(ip_address:int) -> bool:
	pass




#16777217
def main() -> None:
	start_ip = int(ipaddress.IPv4Address(sys.argv[1]))
	end_ip = int(ipaddress.IPv4Address(sys.argv[2]))
	for num in range(start_ip,end_ip):
		flag = True
		for banned in SKIP_LIST:
			if isinstance(banned,range) and num in banned:
				flag = False
				break
			elif num == banned:
				flag = False
				break
		if (flag):
			try:
				print(call_dig(num))
			except:
				print("Failed!")
	


if __name__ == "__main__":
	main()

File: ip_scan.cpp

#include "ip_scan.h"

Query::Query()
{
    this -> packet_for_everyone = make_dns_packet();
    this -> outbound_socket     = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
    this -> inbound_socket      = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);

    if(outbound_socket < 0)
        printf("Create FAILED\n");
}

Query::~Query()
{
    close(outbound_socket);

    close(inbound_socket);
}

Query::packet Query::make_dns_packet()
{
    std::string message = "For research use@Case Western Reserve, contact jxm959@case.edu";

    Query::packet sent_packet = Query::packet();

    return sent_packet;
}

bool Query::is_working(sockaddr_in dns_server) const
{
    return this->send_request(dns_server);
}

bool Query::send_request(sockaddr_in dns_server) const
{
    struct sockaddr_in dest;

    dest.sin_addr.s_addr = inet_addr("172.20.38.75");

    dest.sin_family = AF_INET;

    dest.sin_port = htons(1053);

    if(bind(outbound_socket, (struct sockaddr*)&dest, sizeof(dest)) < 0)
    {

        printf("%s\n", strerror(errno));

        printf("Bind FAILED\n");

        return false;
    }

    if(connect(outbound_socket, (struct sockaddr*)&dns_server, sizeof(dns_server)) < 0)
    {
        printf("Connect FAILED\n");

        return false;
    }

    if(send(outbound_socket, &packet_for_everyone, 400, NULL) < 0)
    {
        printf("Send FAILED\n");

        return false;
    }

    return true and listen_and_check_result();
}

bool Query::listen_and_check_result() const
{
    char recv_buffer[1024];

    recv(outbound_socket, recv_buffer, sizeof(Query::packet), NULL);

    if(strlen(recv_buffer))
        printf("%s\n", recv_buffer);

    return true;
}

int main()
{
    Query query = Query();

    struct sockaddr_in dest;

    dest.sin_addr.s_addr = inet_addr("8.8.8.8");

    dest.sin_family = AF_INET;

    dest.sin_port = htons(53);

    query.is_working(dest);

    return 0;
}

